### PR TITLE
Added `register_multi_post_accumulate_grad_hook`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132610
* #129259
* __->__ #131949

This PR adds `torch.autograd.graph.register_multi_post_accumulate_grad_hook()` that registers a hook that runs after all passed-in leaf tensors' `AccumulateGrad` nodes. In other words, this is like an "all" hook over their `register_post_accumulate_grad_hook()`s.

```
def register_multi_post_accumulate_grad_hook(
    tensors: Sequence[torch.Tensor], fn: Callable[[Tuple[torch.Tensor, ...]], None]
) -> RemovableHandle:
```

The hook `fn` is expected to have the same signature as for `register_post_accumulate_grad_hook`:
```
def fn(tensor: torch.Tensor) -> None:
```

This function can be useful for distributed applications where we want to run some gradient reduction after a group of tensors' gradients have been computed (and accumulated).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/autograd/graph.py [hunk 1]
```diff
@@ -547,6 +547,63 @@ def wrapped_fn(grad: torch.Tensor) -> None:
...
+        )
+    if not all(t.is_leaf for t in tensors):
+        raise ValueError("requires all tensors to be leaf tensors")
+    if len(tensors) == 0:
+        return _MultiHandle(tuple())
+
+    acc_grads = [_get_grad_fn_or_grad_acc(t) for t in tensors]
+    count: Dict[int, int] = dict()
+    nb_calls = None
+    lock = threading.Lock()
```
Comment (awgu) [inline torch/autograd/graph.py:583]: credit: https://github.com/pytorch/pytorch/pull/132055

File: torch/autograd/graph.py [hunk 2]
```diff
@@ -547,6 +547,63 @@ def wrapped_fn(grad: torch.Tensor) -> None:
...
+    and the hook will not wait for its gradient to be accumulated.
+
+    The hook ``fn`` should have the following signature::
+
+        fn(tensors: Tuple[torch.Tensor, ...]) -> None
+
+    The hook ``fn`` will get the ``tensors`` passed as :attr:`tensors`. The
+    hook is like an "all" version of :meth:`torch.Tensor.register_post_accumulate_grad_hook`,
+    where the hook only runs after each tensor's post-accumulate-grad hook
+    would have run.
```
Comment (soulitzer) [inline torch/autograd/graph.py:569]:
```suggestion
    has run.
```

File: torch/autograd/graph.py [hunk 3]
```diff
@@ -547,6 +547,63 @@ def wrapped_fn(grad: torch.Tensor) -> None:
...
+    not needed to compute the gradients for any ``inputs`` specified for the
+    current ``.backward()`` or ``.grad()`` call, this tensor will be ignored,
+    and the hook will not wait for its gradient to be accumulated.
+
+    The hook ``fn`` should have the following signature::
+
+        fn(tensors: Tuple[torch.Tensor, ...]) -> None
+
+    The hook ``fn`` will get the ``tensors`` passed as :attr:`tensors`. The
+    hook is like an "all" version of :meth:`torch.Tensor.register_post_accumulate_grad_hook`,
```
Comment (soulitzer) [inline torch/autograd/graph.py:567]:
```suggestion
    hook is like an "all" multi-hook version of :meth:`torch.Tensor.register_post_accumulate_grad_hook`,
```

**Review by soulitzer — APPROVED** [review]
Thanks, looks great! Just some minor comments on the doc wording

**Review by sanketpurandare — APPROVED** [review]
Much awaited feature :)

**Comment by awgu** [comment]
going to wait for the FSDP2 PR on top to fix the compile issue before landing this one

**Pull request closed without merging**